aten/src/ATen/native/cuda/Assert.cuh

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <c10/cuda/CUDAAssert.h>
+
+namespace at { namespace native { namespace assert {
+
+using namespace c10::cuda;
+
+static inline C10_HOST_DEVICE char* device_strcpy(char* dst, const char* src, int n){
+  int i = 0;
+  while (i+1 < n && src[i]) {
+    dst[i] = src[i];
+    ++i;
+  }
+  if (i < n) {
+    dst[i] = '\0';
+  }
+  return dst;
+}
+
+// If the assert error state was not set before this function returns true, otherwise false.
+static inline C10_HOST_DEVICE bool assert_failed(CUDAAssert* assert_state, CUDAAssertKind type, const char* message, uint32_t line, const char* file) {
+#ifdef __CUDA_ARCH__
+  if (atomicCAS(const_cast<int32_t*>(&assert_state->error), 0, 1) != 0) {
+    return false;
+  }
+#else
+  assert(false);    // should never be called
+#endif
+
+  // copy message and type
+  device_strcpy(assert_state->message, message, MAX_ASSERT_MSG_LENGTH);
+  assert_state->type = type;
+
+  // store call site information
+  device_strcpy(assert_state->file, file, MAX_ASSERT_MSG_LENGTH);
+  assert_state->line = line;
+
+  return true;
+}
+
+static inline C10_DEVICE bool graceful_assert_(CUDAAssert* assert_state, bool condition, const char* message, uint32_t line, const char* file) {
+  if (!condition) {
+    assert_failed(assert_state, CUDAAssertKind::ASSERTION_FAILED, message, line, file);
+  }
+
+  return !assert_state->error;    // return false if we are in error state, signals kernel to quit
```

Comment:
I guess it will be good when you do your performance testing, but I wonder what the perf implication of this atomic access is. Surely it can't be good...